### PR TITLE
Issue #108: Shortcuts for boolean arguments

### PR DESCRIPTION
**Pull request by alexshadley**
In response to [this issue](https://github.com/google/python-fire/issues/108).  This adds a feature where a shortcut can be used for boolean arguments of functions.  For example the function

`def foo(bar=False):`

can be invoked as `foo -b` to set bar to True.  While developing this, I came across a couple concerns:

- In detecting whether or not arguments matched the pattern of a bool shortcut, I decided that python's regex library was the best way to solve this problem.  Is adding `re` to imports ok?
- If a naming collision occurs (see final assert of `testBoolShortcutParsing`) the program raises a FireError describing the ambiguity.  Is this a proper manner to use this error in?
- Currently, if a shortcut doesn't match any function argument, it is parsed as-is.  Should something else be done in this scenario?

### CODE REVIEW

**Comment by dbieber** [comment]
tl;dr I think we should support single-character flags (like your PR does), but not restrict them to bool-only values.

----

I was previously toying with the idea of allowing users to specify a flag (bool or otherwise) that is a prefix of exactly one valid parameter, but ultimately decided against that idea. Here is my thinking there (stronger points in bold):

Pros: 
- **This enables single-character flags**
- And is additionally even more general, handling the case where multiple params start with the same letter.

Cons:
- It enables people to accidentally pass args without knowing precisely what arg they're passing
- **Backwards compatibility for Fire CLIs**
  - For example, if there's a Fire CLI with param send_message, a user might make a script using that CLI using the flag `--send`. If the CLI author later wants to add a new param send_attachment, then the user's script becomes ambiguous and breaks.
- Backwards compatibility for Fire itself


-------

Your proposal differs in two significant ways: single-character-only instead of arbitrary prefixes, and bool-only instead of any type.

Single-character only still gets the most important Pro, which is supporting single-character flags.
Single-character only still has the most important Con too, which is that a single-character flag may be invalidated later.
Single-character only doesn't have as much of the final Con (backwards compatibility for Fire) since it promises a strict subset of what arbitrary-prefix does to Fire's users.
Similarly bool-only promises a strict subset of what any-type does.
However, bool-only deviates from one of Fire's current properties which is that you can pass any value (literal/container of literals) for any argument.


So my thought is that we should support single-character flags, but not restrict them to bool-only.

File: fire/core.py [hunk 1]
```diff
@@ -852,6 +852,21 @@ def _ParseKeywordArgs(args, fn_spec):
...
+  return _IsSingleCharFlag(argument) or _IsMultiCharFlag(argument)
+
+
+def _IsSingleCharFlag(argument):
+  """Determines if the argument is a single char flag (e.g. '-a')"""
+  return re.match('^-[a-z]$', argument)
+
+
+def _IsMultiCharFlag(argument):
+  """Determines if the argument is a multi char flag (e.g. '-alpha')"""
```
Comment (dbieber) [inline fire/core.py]: I think you meant '--alpha'

Comment (alexshadley) [reply fire/core.py]: Yup!  Just pushed a fix

**Comment by dbieber** [comment]
Oh, looks like you're actively working on this (excellent!), I'll hold off on looking through until you're settled.
If you're put off by having to go to travis to check the lint errors (you get there by clicking the red x next to your commit) I can clean those up on my end instead, just lmk.

File: fire/core.py [hunk 2]
```diff
@@ -785,15 +789,27 @@ def _ParseKeywordArgs(args, fn_spec):
...
       got_argument = False
 
-      keyword = argument[2:]
+      keyword = ''
+      if _IsSingleCharFlag(argument):
+        keychar = argument[1]
+        potential_args = [arg for arg in fn_args if arg[0] == keychar]
+        if len(potential_args) == 1:
+          keyword = potential_args[0]
+        elif len(potential_args) > 1:
```
Comment (dbieber) [inline fire/core.py]: we also need to handle the len(potential_args) == 0 case.

Comment (alexshadley) [reply fire/core.py]: Should we also throw an error in this case?  My original assumption here was that if the flag doesn't match, we would assume it was an value for another flag.  Now that I think about it though this could cause confusing behavior.

Comment (dbieber) [reply fire/core.py]: Yeah I think we should throw an error in this case.
If the argument looks like a flag (_IsFlag), it shouldn't be acceptable as a value.
[And I think we'll also broaden what _IsFlag accepts too in a separate change.]

Comment (alexshadley) [reply fire/core.py]: Ok that sounds good to me, I'll add a branch and a test case for it

**Comment by alexshadley** [comment]
Alright, I've run out of time tonight to work on this tonight.  I'll have time tomorrow night to finish up!

**Comment by dbieber** [comment]
I'm going to squash and merge what you have so far, and you can put subsequent improvements in a new PR.

**Comment by dbieber** [comment]
Some next steps:
- raise a FireError if the arg doesn't match any keyword
- give exact matches priority (e.g. if there are args named "a" and "apple" then -a should match "a", rather than being flagged as ambiguous)
- support equal signs with single-character args (-a=10 should be OK)

**Comment by dbieber** [comment]
Let me know if you're interested in taking on these and if so, which ones

**Comment by dbieber** [comment]
> raise a FireError if the arg doesn't match any keyword
Ah, looking back I see that's actually not how things are currently handled. Unparsed args are returned as part of remaining_kwargs.

**Comment by alexshadley** [comment]
Yeah, I can start working on supporting equal sings in single-char args.

As far as raising a FireError if the arg doesn't match, how should we proceed?  I agree that it would be a nice feature, but as I looked into it further this seems to get tricky.  As far as I understand (correct me if I'm wrong here), we first attempt to instantiate the owning class with the arglist, and upon this failing we move on to calling the function.  When I attempted to raise an error on no applicable shortcuts, the error was just thrown whenever the class was parsed as the class had no args at all.

**Comment by dbieber** [comment]
Ah was just about to post here, 1 sec:

**Comment by dbieber** [comment]
I went ahead and put together a commit to add support for equal signs and giving exact matches priority already, I'll put it on GitHub now and drop a link here.

**Comment by dbieber** [comment]
As for raising the FireError, I was mistaken about that. The correct behavior (e.g. in order to support --help properly) is to return unused kwargs in the remaining_kwargs entry of _ParseKeywordArgs. I do this in my commit.

**Comment by dbieber** [comment]
https://github.com/google/python-fire/commit/0b0f2421ea8edcc11a80aadb1f574d687e94664f

**Comment by alexshadley** [comment]
Ah, looks good then!  Any other areas of the project that need work done?

**Comment by dbieber** [comment]
https://github.com/google/python-fire/issues/102 / https://github.com/google/python-fire/issues/97 looks like a fun one if you're up for AST traversals.
